File: src/main/java/top/chiloven/lukosbot2/commands/impl/translate/TranslateCommand.java
package top.chiloven.lukosbot2.commands.impl.translate;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.arguments.StringArgumentType;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Service;
import top.chiloven.lukosbot2.commands.IBotCommand;
import top.chiloven.lukosbot2.commands.UsageNode;
import top.chiloven.lukosbot2.config.CommandConfigProp;
import top.chiloven.lukosbot2.config.CommandConfigProp.Translate;
import top.chiloven.lukosbot2.core.command.CommandSource;

import java.util.Arrays;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import static top.chiloven.lukosbot2.util.brigadier.builder.LiteralArgumentBuilder.literal;
import static top.chiloven.lukosbot2.util.brigadier.builder.RequiredArgumentBuilder.argument;

@Service
@ConditionalOnProperty(
        prefix = "lukos.commands.control",
        name = "translate",
        havingValue = "true",
        matchIfMissing = true
)
public class TranslateCommand implements IBotCommand {
    public final TranslationService ts;
    private final Translate translate;

    public TranslateCommand(CommandConfigProp ccp) {
        this.translate = ccp.getTranslate();
        this.ts = new TranslationService(translate);
    }

    @Override
    public String name() {
        return "translate";
    }

    @Override
    public String description() {
        return "";
    }

    @Override
    public UsageNode usage() {
        return UsageNode.root(name())
                .description(description())
                .syntax("翻译文本（默认自动检测语言）", UsageNode.arg("text"))
                .syntax("翻译文本（可选指定源/目标语言）",
                        UsageNode.opt(UsageNode.group(UsageNode.lit("-f"), UsageNode.arg("from_lang"))),
                        UsageNode.opt(UsageNode.group(UsageNode.lit("-t"), UsageNode.arg("to_lang"))),
                        UsageNode.arg("text")
                )
                .param("text", "要翻译的文本（支持空格）")
                .option("-f", "指定源语言（可选；不指定则自动检测）")
                .option("-t", "指定目标语言（可选；不指定则使用默认目标语言）")
                .example(
                        "/translate the quick brown fox over the lazy dog",
                        "/translate -f en -t zh-Hans the quick brown fox over the lazy dog",
                        "/translate -t ja Do not go gentle into that good night"
                )
                .build();
    }

    @Override
    public void register(CommandDispatcher<CommandSource> dispatcher) {
        dispatcher.register(
                literal(name())
                        .executes(ctx -> {
                            ctx.getSource().reply(usageText());
                            return 0;
                        })

                        .then(argument("args", StringArgumentType.greedyString())
                                .executes(ctx -> {
                                    try {
                                        CommandSource src = ctx.getSource();
                                        String args = StringArgumentType.getString(ctx, "args");

                                        var results = Pattern.compile("-f\\s+(?<from>\\S+)|-t\\s+(?<to>\\S+)")
                                                .matcher(args)
                                                .results()
                                                .toList();

                                        src.reply(run(
                                                results.stream()
                                                        .map(r -> r.group("from"))
                                                        .findFirst()
                                                        .orElse("auto"),
                                                results.stream()
                                                        .map(r1 -> r1.group("to"))
                                                        .findFirst()
                                                        .orElse(translate.getDefaultLang()),
                                                Arrays.stream(args.split("\\s+"))
                                                        .filter(s -> !s.equals("-f") && !s.equals("-t"))
                                                        .filter(s -> results.stream().noneMatch(r2 ->
                                                                s.equals(r2.group("from")) || s.equals(r2.group("to"))
                                                        ))
                                                        .collect(Collectors.joining(" "))
                                                        .trim()));
                                        return 0;
                                    } catch (Exception e) {
                                        ctx.getSource().reply("不正确的格式。");
                                        return 1;
                                    }
                                })
                        )
        );
    }

    private String run(String from, String to, String text) {
        return ts.translate(from, to, text);
    }
}

File: src/main/java/top/chiloven/lukosbot2/commands/bot/translate/TranslateCommand.kt
package top.chiloven.lukosbot2.commands.bot.translate

import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.stereotype.Service
import top.chiloven.lukosbot2.commands.IBotCommand
import top.chiloven.lukosbot2.config.CommandConfigProp
import top.chiloven.lukosbot2.core.command.definition.ArgType
import top.chiloven.lukosbot2.core.command.definition.dsl.*

@Service
@ConditionalOnProperty(
    prefix = "lukos.commands.control",
    name = ["translate"],
    havingValue = "true",
    matchIfMissing = true
)
class TranslateCommand(ccp: CommandConfigProp) : IBotCommand {

    private val translate = ccp.translate
    private val ts = TranslationService(translate)

    override fun definition() = botCommand("translate") {
        alias("tr")
        description = "翻译文本"

        argv {
            positional("text", ArgType.StringType) { required = true; greedy = true; description = "要翻译的文本" }
            option("from") { names = listOf("-f"); type = ArgType.StringType; description = "源语言" }
            option("to") { names = listOf("-t"); type = ArgType.StringType; description = "目标语言" }
            execute { args ->
                val from = args.getOrNull<String>("from") ?: "auto"
                val to = args.getOrNull<String>("to") ?: translate.defaultLang
                val text = args.get<String>("text")
                source.reply(ts.translate(from, to, text))
            }
        }

        syntax(
            "翻译文本（可选指定源/目标语言）",
            opt(group(lit("-f"), arg("from_lang"))),
            opt(group(lit("-t"), arg("to_lang"))),
            arg("text")
        )
        param("text", "要翻译的文本（支持空格）")
        optionDoc("-f", "指定源语言（可选；不指定则自动检测）")
        optionDoc("-t", "指定目标语言（可选；不指定则使用默认目标语言）")
        example(
            "/translate the quick brown fox",
            "/translate -f en -t zh-Hans the quick brown fox",
            "/translate -t ja Hello"
        )
    }

}
